compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/HeaderProcessor.kt

```diff
@@ -121,58 +128,67 @@ internal class ProgressionLoopHeader(
-            //   var last = B
-            //   if (inductionVar <= last) {
-            //     // Loop is not empty
-            //     do {
-            //       val loopVar = inductionVar
-            //       inductionVar++
-            //       // Loop body
-            //     } while (loopVar != last)
-            //   }
-            //
-            // The `if (inductionVar <= last)` "not empty" check is added in buildNotEmptyConditionIfNecessary().
-            assert(loopVariable != null)
-            val booleanNotFun = context.irBuiltIns.booleanClass.functions.first { it.owner.name.asString() == "not" }
-            val newCondition = irCallOp(booleanNotFun, booleanNotFun.owner.returnType, irCall(context.irBuiltIns.eqeqSymbol).apply {
-                putValueArgument(0, irGet(loopVariable!!))
-                putValueArgument(1, irGet(last))
-            })
-            IrDoWhileLoopImpl(loop.startOffset, loop.endOffset, loop.type, loop.origin).apply {
-                label = loop.label
-                condition = newCondition
-                body = newBody
+    override fun buildLoop(builder: DeclarationIrBuilder, oldLoop: IrLoop, newBody: IrExpression?): LoopReplacement {
+        with(builder) {
+            var (newLoop, replacementExpression) = if (headerInfo.canOverflow) {
+                // If the induction variable CAN overflow, we cannot use it in the loop condition. Loop is lowered into something like:
+                //
+                //   if (inductionVar <= last) {
+                //     // Loop is not empty
+                //     do {
+                //       val loopVar = inductionVar
+                //       inductionVar += step
+                //       // Loop body
+                //     } while (loopVar != last)
+                //   }
+                assert(loopVariable != null)
+                val booleanNotFun = context.irBuiltIns.booleanClass.functions.first { it.owner.name.asString() == "not" }
+                val newCondition = irCallOp(booleanNotFun, booleanNotFun.owner.returnType, irCall(context.irBuiltIns.eqeqSymbol).apply {
+                    putValueArgument(0, irGet(loopVariable!!))
+                    putValueArgument(1, irGet(last))
+                })
+                val newLoop = IrDoWhileLoopImpl(oldLoop.startOffset, oldLoop.endOffset, oldLoop.type, oldLoop.origin).apply {
+                    label = oldLoop.label
+                    condition = newCondition
+                    body = newBody
+                }
+                val notEmptyCheck = irIfThen(buildLoopCondition(builder), newLoop)
+                LoopReplacement(newLoop, notEmptyCheck)
+            } else {
+                // If the induction variable can NOT overflow, use a simple while loop. Loop is lowered into something like:
+                //
+                //   while (inductionVar <= last) {
+                //       val loopVar = inductionVar
+                //       inductionVar += step
```

Comment:
Why not 
```
                //   while (loopVar <= last) {
                //       loopVar = inductionVar
                //       // Loop body
                //       loopVar += step                
                //   }
```

Reply:
That would mean moving `loopVar` declaration outside of the while loop, and removing `inductionVar` since it's not needed anymore. It's possible, but it would only work just for this one case.

Reply:
As I see that could affect reversing